tests/test_ethash.py

```diff
@@ -0,0 +1,245 @@
+        == b"\x00" * 32
+    )
+    assert (
+        generate_seed(Uint(EPOCH_SIZE))
+        == generate_seed(Uint(2 * EPOCH_SIZE - 1))
+        == keccak256(b"\x00" * 32)
+    )
+    # NOTE: The below bytes value was obtained by obtaining the seed for the same block number from Geth.
+    assert (
+        generate_seed(Uint(12345678))
+        == b"[\x8c\xa5\xaaC\x05\xae\xed<\x87\x1d\xbc\xabQBGj\xfd;\x9cJ\x98\xf6Dq\\z\xaao\x1c\xf7\x03"
+    )
+
+
+def test_epoch_start_and_end_blocks_have_same_seed() -> None:
+    for _ in range(100):
+        block_number = Uint(randint(10000, 20000))
+        epoch_start_block_number = (block_number // EPOCH_SIZE) * EPOCH_SIZE
+        epoch_end_block_number = epoch_start_block_number + EPOCH_SIZE - 1
+
+        assert (
+            generate_seed(epoch_start_block_number)
+            == generate_seed(block_number)
+            == generate_seed(epoch_end_block_number)
+        )
+
+
+#
+# Geth DAG related functionalities for fuzz testing
+#
+
+
+def download_geth(dir: str) -> None:
+    geth_release_name = "geth-linux-amd64-1.10.8-26675454"
+    # 26 seconds to fetch Geth. 1.5 minute for each epoch dataset creation
+    url = f"https://gethstore.blob.core.windows.net/builds/{geth_release_name}.tar.gz"
+    r = requests.get(url)
+
+    with open(f"{dir}/geth.tar.gz", "wb") as f:
+        f.write(r.content)
+
+    geth_tar = tarfile.open(f"{dir}/geth.tar.gz")
+    geth_tar.extractall(dir)
+
+    shutil.move(f"{dir}/{geth_release_name}/geth", dir)
+    shutil.rmtree(f"{dir}/{geth_release_name}", ignore_errors=True)
+    os.remove(f"{dir}/geth.tar.gz")
+
+
+def generate_dag_via_geth(
+    geth_path: str, block_number: Uint, dag_dump_dir: str
+) -> None:
+    subprocess.call([geth_path, "makedag", str(block_number), dag_dump_dir])
+
+
+def fetch_dag_data(dag_dump_dir: str, epoch_seed: bytes) -> Tuple[bytes, ...]:
+    dag_file_path = f"{dag_dump_dir}/full-R23-{epoch_seed.hex()[:16]}"
+    with open(dag_file_path, "rb") as fp:
+        dag_dataset = fp.read()
+        # The first 8 bytes are Magic Bytes and can be ignored.
+        dag_dataset = dag_dataset[8:]
+
+    dag_dataset_items = []
+    for i in range(0, len(dag_dataset), HASH_BYTES):
+        dag_dataset_items.append(dag_dataset[i : i + HASH_BYTES])
+
+    return tuple(dag_dataset_items)
+
+
+def test_dataset_generation_random_epoch(tmpdir: str) -> None:
+    """
+    Generate a random epoch and obtain the DAG for that epoch from geth.
+    Then ensure the following 2 test scenarios:
+        1. The first 100 dataset indices are same when the python
+        implementation is compared with the DAG dataset.
+        2. Randomly take 500 indices between
+        [101, `dataset size in words` - 1] and ensure that the values are
+        same between python implementation and DAG dataset.
+    """
+    download_geth(tmpdir)
+
+    epoch_number = Uint(randint(0, 100))
+    block_number = epoch_number * HASH_BYTES
```

Comment:
I think this should be `block_number = epoch_number * EPOCH_SIZE`.

Reply:
Not really. I kind of used `HASH_BYTES` instead of `EPOCH_SIZE` as I felt it would add a bit more entropy rather than testing on only the end blocks of the epoch.

Now to think of it, it makes sense to actually make this block_number random as well within the epoch for much more entropy. I'll be pushing this change in a while.